doc/user-manual/language/type-based-termination-checking.rst

```diff
@@ -0,0 +1,317 @@
+
+::
+
+      div : Nat → Nat → Nat
+      div zero    y = zero
+      div (suc x) y = suc (div (minus x y) y)
+
+Another interesting application of size preservation can be found in combination with coinductive functions. For coinduction, size preservation seeks to determine whether it is possible to assign a fixed *codomain* size to some of the *domain* sizes. In other words, inductive definitions can be size-preserving in their output, while coinductive definitions can be size-preserving in their input.
+
+For example, consider a coinductive function ``zipWith``:
+
+::
+
+    zipWith : (Nat → Nat → Nat) → Stream → Stream → Stream
+    zipWith f s1 s2 .head = f (s1 .head) (s2 .head)
+    zipWith f s1 s2 .tail = zipWith f (s1 .tail) (s2 .tail)
+
+Here, the depth of the returned ``Stream`` is the same as the requested depth of incoming ``s1`` and ``s2``. The type-based termination checker recognizes this, concluding that all three ``s1``, ``s2``, and the returned stream share the same size variable.
+
+Given size-preserving ``zipWith``, Agda is able to define an infinite stream of Fibonacci numbers:
+
+::
+
+   fib : Stream
+   fib .head = zero
+   fib .tail .head = suc zero
+   fib .tail .tail = zipWith plus fib (fib .tail)
+
+This function passes termination checking. We shall explain the logic of Agda for the third clause.
+
+Following our intuition with coinductive functions, the are three depth parameters ``k < m < n``, where the outer stream is of depth ``n``, and to pass checking the third clause should return the stream of depth at least ``k``. If the first inner ``fib`` is used with the depth ``k``, and the second ``fib`` is used with the depth ``m`` (note, that the smallest available depth for ``fib .tail`` is ``k``, hence ``fib`` must use something bigger, which is ``m``), then the size-preserving ``zipWith`` returns a stream of size ``k``, which is indeed what is required from it. Now we see that both recursive calls to ``fib`` are performed with depths ``k`` and ``m``, which are smaller than ``n``. Agda concludes that this function is terminating.
+
+Size preservation is tightly coupled with polarities. Given a function signature, all occurrences of *negative inductive* and *positive coinductive* variables are considered as input, and they serve as possible candidates for size preservation analysis. On the other hand, all *positive inductive* and *negative coinductive* variables are considered as output, and a function signature may be size-preserving precisely in them. For example, consider the following definition:
+
+::
+
+    foo : {A : Set} → (Nat → A) → Nat → A
+    foo f x = f x
+
+
+
+Here, the first ``Nat`` in ``foo`` is in a doubly-negative position, which means that the position is positive, and ``foo`` can be size-preserving in the first ``Nat``. From the definition we see that it is indeed the case. One application of this fact is that the following function passes termination check:
```

Comment:
What does it even mean for `foo` to be "size-preserving in the first `Nat`"? It does not produce a `Nat` as an output, so how could it preserve the size of it?

Reply:
This function does produce `Nat` as an output, the only weird thing here is that the output is in a positive position, not in a strictly positive one. As a user of `foo`, you can obtain this first `Nat`, as it is supplied by `foo`. On the other hand, the second `Nat` has to be supplied by _you_, and it is `foo` which obtains it.